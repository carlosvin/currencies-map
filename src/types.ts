export declare type Display = 'symbol' | 'name';
