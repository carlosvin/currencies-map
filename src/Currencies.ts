import codes from './codes';

export class Currencies extends Map<string, string> {

    private static readonly RE = /(\d|,|\.)+/;
    private static readonly NUM = Number(0);

    /** 
     * @param currencyDisplay Configure to show currency symbol or name 
     * */
    constructor(currencyDisplay: 'symbol' | 'name') {
        super();
        codes.forEach(c => this.set(c, Currencies._format(c, currencyDisplay)));
    }

    private static _format(currency: string, currencyDisplay: 'symbol' | 'name') {
        const currencyNumber = Currencies.NUM.toLocaleString(
            navigator ? navigator.language : undefined, 
            { style: 'currency', currencyDisplay, currency });
        return currencyNumber.replace(Currencies.RE, '').trim();
    }
}
