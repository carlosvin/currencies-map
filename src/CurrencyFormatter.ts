import { Display } from "./types";

export class CurrencyFormatter {
    private static readonly RE = /(\d|,|\.)+/;
    readonly formatter: Intl.NumberFormat;
    readonly currency: string;

    constructor(currencyDisplay: Display, currency: string, locales?: string[]) {
        this.formatter = new Intl.NumberFormat(
            locales, { style: 'currency', currencyDisplay, currency });
        this.currency = currency;
    }

    get text() {
        return this.formatter.format(0).replace(CurrencyFormatter.RE, '').trim();
    }
}
