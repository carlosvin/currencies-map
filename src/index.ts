/** List with currency ISO codes */
const  CODES  = ["BDT", "EUR", "XOF", "BGN", "BAM", "BBD", "XPF", "BMD", "BND", "BOB", "BHD", "BIF", "BTN", "JMD", "BWP", "WST", "BRL", "BSD", "BYR", "BZD", "RUB", "RWF", "RSD", "USD", "TMT", "TJS", "RON", "NZD", "GTQ", "GBP", "XAF", "JPY", "GYD", "GEL", "GNF", "GMD", "GIP", "GHS", "OMR", "TND", "JOD", "HRK", "HTG", "HUF", "HKD", "HNL", "AUD", "VEF", "NOK", "PYG", "IQD", "PAB", "PGK", "PEN", "PKR", "PHP", "PLN", "ZMK", "EGP", "ZAR", "VND", "SBD", "ETB", "SOS", "ZWL", "SAR", "ERN", "MDL", "MGA", "MAD", "UZS", "MMK", "MOP", "MNT", "MKD", "MUR", "MWK", "MVR", "XCD", "MRO", "UGX", "TZS", "MYR", "MXN", "ILS", "SHP", "FJD", "FKP", "DKK", "NIO", "NAD", "VUV", "NGN", "NPR", "CHF", "COP", "CNY", "CLP", "CAD", "CDF", "CZK", "CRC", "ANG", "CVE", "CUP", "SZL", "SYP", "KGS", "KES", "SSP", "SRD", "KHR", "KMF", "STD", "KRW", "KPW", "KWD", "SLL", "SCR", "KZT", "KYD", "SGD", "SEK", "SDG", "DOP", "DJF", "YER", "DZD", "UYU", "LBP", "LAK", "TWD", "TTD", "TRY", "LKR", "TOP", "LTL", "LRD", "LSL", "THB", "LYD", "AED", "AFN", "ISK", "IRR", "AMD", "ALL", "AOA", "ARS", "AWG", "INR", "AZN", "IDR", "UAH", "QAR", "MZN"];

class Currencies extends Map<string, string> {

    private static readonly RE = /(\d|,|\.)+/;
    private static readonly NUM = Number(0);

    /** 
     * @param currencyDisplay Configure to show currency symbol or name 
     * */
    constructor(currencyDisplay: 'symbol' | 'name') {
        super();
        CODES.forEach(c => this.set(c, Currencies._format(c, currencyDisplay)));
    }

    private static _format(currency: string, currencyDisplay: 'symbol' | 'name') {
        const currencyNumber = Currencies.NUM.toLocaleString(
            navigator ? navigator.language : undefined, 
            { style: 'currency', currencyDisplay, currency });
        return currencyNumber.replace(Currencies.RE, '').trim();
    }

    private static _names: Currencies;

    static get names () : Currencies {
        if (!this._names) {
            this._names = new Currencies('name');
        }
        return this._names;
    }

    private static _symbols: Currencies;

    static get symbols () : Currencies {
        if (!this._symbols) {
            this._symbols = new Currencies('symbol');
        }
        return this._symbols;
    }
}
export { Currencies, CODES };
