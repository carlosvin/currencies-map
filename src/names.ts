import { Currencies } from "./Currencies";

export const namesMap = new Currencies('name');
export default namesMap;
