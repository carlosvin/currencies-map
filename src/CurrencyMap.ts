import { CurrencyFormatter } from "./CurrencyFormatter";
import { CODES } from "./codes";
import { Display } from "./types";

/** 
 * Map from currency codes to currency names or symbols
 */
export class CurrencyMap extends Map<string, string> {

    /** 
     * @param currencyDisplay Configure to show currency symbol or name 
     * */
    constructor(currencyDisplay: Display, locales?: string[]) {
        super(CODES
            .map<[string, string]>(c => [c, new CurrencyFormatter(currencyDisplay, c, locales).text])
            .sort((a, b) => a[1].localeCompare(b[1])));
    }
}
