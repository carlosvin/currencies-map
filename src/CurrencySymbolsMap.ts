import { CurrencyMap } from "./CurrencyMap";

const currenciesSymbolsMap = new CurrencyMap('symbol');
export default currenciesSymbolsMap;
