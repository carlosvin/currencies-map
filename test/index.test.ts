import { Currencies } from '../src/index';

let languageGetter: any;

beforeEach(() => {
    languageGetter = jest.spyOn(window.navigator, 'languages', 'get');
    languageGetter.mockReturnValue(['es-ES']);    
})

test('Load names singleton', async () => {
    expect(Currencies.names.get('THB')).toStrictEqual('Thai baht');
});

test('Load symbols singleton', async () => {
    expect(Currencies.symbols.get('EUR')).toStrictEqual('€');
});

test('Euro symbol en-GB', () => {
    languageGetter.mockReturnValue();
    const cs = new Currencies('symbol', ['es-ES']);
    expect(cs.get('EUR')).toStrictEqual('€');
});

test('Myanmar kyats name es-ES', () => {
    const cs = new Currencies('name', ['es-ES']);
    expect(cs.get('MMK')).toStrictEqual('kiats');
});

test('Myanmar kyats name en-US', () => {
    const cs = new Currencies('name', ['en-US']);
    expect(cs.get('MMK')).toStrictEqual('Myanmar kyats');
});

test('USD name and symbol es-ES', () => {
    expect(new Currencies('name', ['es-ES']).get('USD')).toStrictEqual('dólares estadounidenses');
    expect(new Currencies('symbol', ['es-ES']).get('USD')).toStrictEqual('US$');
});

test('ARS name and symbol es-AR', () => {
    expect(new Currencies('name', ['es-AR']).get('ARS')).toStrictEqual('pesos argentinos');
    expect(new Currencies('symbol', ['es-AR']).get('ARS')).toStrictEqual('$');
});

test('Load all names', () => {
    const observed = [...new Currencies('name', ['es-ES']).values()];
    expect(observed).toMatchSnapshot();
});

test('Load all symbols', () => {
    const observed = [...new Currencies('symbol', ['es-ES']).values()];
    expect(observed).toMatchSnapshot();
});

test('Load defaults', () => {
    const observed = Currencies.names.loadDefaults([['YYY', 'YYY value']]);
    expect(observed.get('YYY')).toBe('YYY value');
    expect(observed.get('EUR')).toBe('euros');
});
