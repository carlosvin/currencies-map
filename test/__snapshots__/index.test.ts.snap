// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Load all names 1`] = `
Array [
  "(moneda desconocida)",
  "afganis",
  "ariaris",
  "balboas panameños",
  "bats",
  "bires",
  "bolívares venezolanos",
  "bolivianos",
  "cedis",
  "chelines kenianos",
  "chelines somalíes",
  "chelines tanzanos",
  "chelines ugandeses",
  "código reservado para pruebas",
  "colones costarricenses",
  "colones salvadoreños",
  "córdobas oro",
  "coronas checas",
  "coronas danesas",
  "coronas islandesas",
  "coronas noruegas",
  "coronas suecas",
  "dalasis",
  "derechos especiales de giro",
  "dinares argelinos",
  "dinares bareiníes",
  "dinares iraquíes",
  "dinares jordanos",
  "dinares kuwaitíes",
  "dinares libios",
  "dinares macedonios",
  "dinares serbios",
  "dinares tunecinos",
  "dírhams de los Emiratos Árabes Unidos",
  "dírhams marroquíes",
  "dobras",
  "dólar zimbabuense",
  "dólares australianos",
  "dólares bahameños",
  "dólares barbadenses",
  "dólares beliceños",
  "dólares bermudeños",
  "dólares bruneanos",
  "dólares canadienses",
  "dólares de las Islas Caimán",
  "dólares de Trinidad y Tobago",
  "dólares del Caribe Oriental",
  "dólares estadounidenses",
  "dólares estadounidenses (día siguiente)",
  "dólares fiyianos",
  "dólares guyaneses",
  "dólares hongkoneses",
  "dólares jamaicanos",
  "dólares liberianos",
  "dólares namibios",
  "dólares neozelandeses",
  "dólares salomonenses",
  "dólares singapurenses",
  "dólares surinameses",
  "dongs",
  "drams",
  "escudos de Cabo Verde",
  "eslotis",
  "euros",
  "euros WIR",
  "florines antillanos",
  "florines arubeños",
  "forintos húngaros",
  "francos burundeses",
  "francos CFA de África Central",
  "francos CFA de África Occidental",
  "francos CFP",
  "francos comorenses",
  "francos congoleños",
  "francos guineanos",
  "francos ruandeses",
  "francos suizos",
  "francos WIR",
  "francos yibutianos",
  "grivnas",
  "guaraníes paraguayos",
  "gultrums",
  "gurdes haitianos",
  "kiats",
  "kinas",
  "kips",
  "kuachas malauíes",
  "kuachas zambianos",
  "kuanzas",
  "kunas",
  "laris",
  "leks",
  "lempiras hondureños",
  "leonas",
  "leus moldavos",
  "leus rumanos",
  "levas búlgaras",
  "libras de Santa Elena",
  "libras egipcias",
  "libras esterlinas",
  "libras gibraltareñas",
  "libras libanesas",
  "libras malvinenses",
  "libras sirias",
  "libras sudanesas",
  "libras sursudanesas",
  "lilangenis",
  "liras turcas",
  "lotis lesotenses",
  "manats azerbaiyanos",
  "manats turcomanos",
  "marcos convertibles de Bosnia y Herzegovina",
  "meticales",
  "MVDOL bolivianos",
  "nairas",
  "nakfas",
  "nuevos dólares taiwaneses",
  "nuevos séqueles israelíes",
  "oro",
  "paangas",
  "paladio",
  "patacas de Macao",
  "pesos argentinos",
  "pesos chilenos",
  "pesos colombianos",
  "pesos cubanos",
  "pesos cubanos convertibles",
  "pesos dominicanos",
  "pesos filipinos",
  "pesos mexicanos",
  "pesos uruguayos",
  "pesos uruguayos en unidades indexadas",
  "plata",
  "platino",
  "pulas",
  "quetzales guatemaltecos",
  "rands",
  "reales brasileños",
  "riales cataríes",
  "riales iraníes",
  "riales omaníes",
  "riales saudíes",
  "riales yemeníes",
  "rieles",
  "ringits",
  "rublos bielorrusos",
  "rublos rusos",
  "rufiyas",
  "rupias esrilanquesas",
  "rupias indias",
  "rupias indonesias",
  "rupias mauricianas",
  "rupias nepalíes",
  "rupias pakistaníes",
  "rupias seychellenses",
  "soles peruanos",
  "somonis tayikos",
  "soms",
  "sums",
  "takas",
  "talas",
  "tengues kazajos",
  "tugriks",
  "uguiyas",
  "unidades compuestas europeas",
  "unidades de cuenta europeas (XBC)",
  "unidades de cuenta europeas (XBD)",
  "unidades de fomento chilenas",
  "unidades de inversión (UDI) mexicanas",
  "unidades de valor reales",
  "unidades monetarias europeas",
  "unidades previsionales uruguayos",
  "vatus",
  "VED",
  "wons norcoreanos",
  "wons surcoreanos",
  "XSU",
  "XUA",
  "yenes",
  "yuanes",
]
`;

exports[`Load all symbols 1`] = `
Array [
  "¤",
  "฿",
  "₫",
  "€",
  "AED",
  "AFN",
  "ALL",
  "AMD",
  "ANG",
  "AOA",
  "ARS",
  "AUD",
  "AWG",
  "AZN",
  "BAM",
  "BBD",
  "BDT",
  "BGN",
  "BHD",
  "BIF",
  "BMD",
  "BND",
  "BOB",
  "BOV",
  "BRL",
  "BSD",
  "BTN",
  "BWP",
  "BYN",
  "BZD",
  "CAD",
  "CDF",
  "CFPF",
  "CHE",
  "CHF",
  "CHW",
  "CLF",
  "CLP",
  "CNY",
  "COP",
  "COU",
  "CRC",
  "CUC",
  "CUP",
  "CVE",
  "CZK",
  "DJF",
  "DKK",
  "DOP",
  "DZD",
  "EGP",
  "ERN",
  "ETB",
  "FJD",
  "FKP",
  "GBP",
  "GEL",
  "GHS",
  "GIP",
  "GMD",
  "GNF",
  "GTQ",
  "GYD",
  "HKD",
  "HNL",
  "HRK",
  "HTG",
  "HUF",
  "IDR",
  "ILS",
  "INR",
  "IQD",
  "IRR",
  "ISK",
  "JMD",
  "JOD",
  "JPY",
  "KES",
  "KGS",
  "KHR",
  "KMF",
  "KPW",
  "KRW",
  "KWD",
  "KYD",
  "KZT",
  "LAK",
  "LBP",
  "LKR",
  "LRD",
  "LSL",
  "LYD",
  "MAD",
  "MDL",
  "MGA",
  "MKD",
  "MMK",
  "MNT",
  "MOP",
  "MRU",
  "MUR",
  "MVR",
  "MWK",
  "MXN",
  "MXV",
  "MYR",
  "MZN",
  "NAD",
  "NGN",
  "NIO",
  "NOK",
  "NPR",
  "NZD",
  "OMR",
  "PAB",
  "PEN",
  "PGK",
  "PHP",
  "PKR",
  "PLN",
  "PYG",
  "QAR",
  "RON",
  "RSD",
  "RUB",
  "RWF",
  "SAR",
  "SBD",
  "SCR",
  "SDG",
  "SEK",
  "SGD",
  "SHP",
  "SLL",
  "SOS",
  "SRD",
  "SSP",
  "STN",
  "SVC",
  "SYP",
  "SZL",
  "TJS",
  "TMT",
  "TND",
  "TOP",
  "TRY",
  "TTD",
  "TWD",
  "TZS",
  "UAH",
  "UGX",
  "US$",
  "USN",
  "UYI",
  "UYU",
  "UYW",
  "UZS",
  "VED",
  "VES",
  "VUV",
  "WST",
  "XAF",
  "XAG",
  "XAU",
  "XBA",
  "XBB",
  "XBC",
  "XBD",
  "XCD",
  "XDR",
  "XOF",
  "XPD",
  "XPT",
  "XSU",
  "XTS",
  "XUA",
  "YER",
  "ZAR",
  "ZMW",
  "ZWL",
]
`;
